### ISSUE TITLE
Use a virtual scroller for the Tree view

### ISSUE THREAD

**Issue by kinow** [cylc/cylc-ui#330]
**Describe exactly what you would like to see in an upcoming release**

Handling hundreds or thousands of components being updated can be slow in Vue. It needs to calculate the difference between Virtual DOM and the browser real DOM. It then needs to update the elements as required, which may trigger [reflow ](https://developers.google.com/speed/docs/insights/browser-reflow).

The virtual scrollers (as in Instagram or FaceBook feed, and others use for pagination too) allow developers to have a view area, that defines how many elements are rendered. Or other approaches such as an initial number of items displayed.

For Vue, [vue-virtual-scroller](https://akryum.github.io/vue-virtual-scroller/#/) is the most famous solution.

So with something like this, we would not display all the `TreeItem` elements, only the ones currently visible to the user. Reducing Vue components, elements in the DOM, and browser reflow.

The current blocker for this is the recursion and hierarchy in the Tree/TreeItem components. But in theory I think we should be able to have a single list of `TreeItem` that specify: type, margin from left (i.e. a child element has the same left margin as the parent, plus `Npx`).

**Additional context**

**Pull requests welcome!**

**Comment by kinow** [comment]
Thought about this yesterday in the gym after the meeting. Besides functional components (which prevent states in the `TreeItem` such as collapsed/expanded), so far this is the only other solution I could think of.

I will try to catch up with some fellow developers that are doing some JS work during holidays, and see what they think about this too.

**Comment by kinow** [comment]
Good video about similar solution for React: [Michael Jackson: Rendering with React | js.la January 2016](https://www.youtube.com/watch?v=7S8v8jfLb1Q&feature=youtu.be&t=26m2s)

**Comment by kinow** [comment]
Branch created: https://github.com/kinow/cylc-ui/tree/optimize-tree-component-1

Right now got it to display a Tree view, with all the items, using a flat list.

![Screen Shot 2019-12-11 at 15 38 13-fullpage](https://user-images.githubusercontent.com/304786/70586919-657cb880-1c2d-11ea-8e2a-2b68a3b044fd.png)

We could then just use the `vue-virtual-scroller` and that's that. But first need to figure how to handle expand/collapse with this flat structure.

The state for expanded/collapse was in the component, not in the data. We have now these two options, keep the state in the component, or move to the data.

If we move to the data, that would mean updating all the N children nodes of one node (i.e. collapsing a cycle point would take longer than collapsing a task).

If we leave the state in the component, that would mean that we need to figure out a way to tell vue that all the other components below the collapsed component, with a `child.depth > component.depth` must be hidden. And it must be with a `v-show`, so that the element is not destroyed/recreated, as would be with a `v-if`.

**Comment by kinow** [comment]
Also interesting to look at the screenshot above. It was taken with Firefox, using the feature to take screen shot of the whole page. If you look at the left sidebar/menu, its height is set to my viewport, i.e. visible area.

If I get this working, taking a screenshot would bring less elements, possibly a few more after the viewport limit. An easy way to test if it's working. Note: the mocked workflow adds 40 `TreeItem` components when rendered at the moment.

**Comment by kinow** [comment]
I thought we needed to have a flat list, in order to create the infinite scroll. But found one JS library that implements an infinite scroll for a tree: https://infinite-tree.js.org/index.html

There is a React component too based on this library. Writing something similar to this component in Vue is not too hard (different than implementing the infinite scroll for a tree I guess).

Will have a better look at this library. Its demo includes filter/search too, which we will need later.

**Comment by hjoliver** [comment]
Good find, that seems worth a look.

**Comment by kinow** [comment]
Dissecting the `infinity-tree` library today (good task for a rainy day in Auckland :cloud_with_rain:). Turns out they have a tree and a flat structure. Interestingly my work-in-progress branch was going in the same direction, but in a different way.

But they do not have the code to flat a tree, instead they used a JS library called [`flattree`](https://www.npmjs.com/package/flattree), by the same author. Might be able to use just that library.

>Convert hierarchical tree structure to flat structure. With a flat structure, it allows you to scroll a large tree easily with virtualization.

Note to self: check NPM for libraries before writing code.

**Comment by kinow** [comment]
Branch with `infinite-tree` library and Cylc UI: https://github.com/kinow/cylc-ui/tree/infinite-tree

Successfully created the object `tree` as in their example, using our `workflowTree` object. The expected attributes match (they are `id`, `name`, and `children`, with the same attributes `id, `name`, ...).

Nothing is displayed in the tree component yet. I suspect what's missing is telling the library how to render the nodes. Hopefully after that it will magically display the treeitem as a simple hierarchical tree, but with the infinite scroll.

If that indeed works, then I can start playing with the complex workflow, and measuring the performance. Assuming it's good enough, then I will just grab the old styles for nodes/tasks/jobs/etc from the `TreeItem`, and move to the new `infinite-tree` render function 🤞

**Comment by kinow** [comment]
Good progress today, got most of the old features working again, but using a mix of

- `infinite-tree`
- `vue-virtual-scroller`
- Our old `TreeItem` component

![image](https://user-images.githubusercontent.com/304786/71329782-a8278480-258d-11ea-9239-62e3dbe60870.png)

The problem I am facing now is that the current version in NPM registry of `infinite-tree` uses an older version of `flattree`. When the flat tree is created from our data (`workflowTree`), the node state is ignored.

So for our current tree, I modified the state to include the `open` flag (used to be called `isExtended` is our old component, but `infinite-tree` has other properties grouped under `node { state {...}}`). The new version of `flattree` handles the state correctly. The `master` branch of `infinite-tree` uses this new version, but it hasn't been released.

I tried installing from the git tag, as with pip, which works (did that before for other NIWA projects). But this time it keeps complaining that it cannot find the `infinite-tree` module, even though it has been downloaded into the `node_modules` folder. So will need to play with NPM and `package.json` to see what's going on.

But we are getting closer :tada:
